### PR TITLE
[CALCITE-3880] Add SortExchange support to RelFieldTrimmer

### PR DESCRIPTION
**Pull request by kasakrisz**
Pull request for https://issues.apache.org/jira/browse/CALCITE-3880

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/RelBuilderTest.java [hunk 1]
```diff
@@ -3362,4 +3364,89 @@ private void checkExpandTable(RelBuilder builder, Matcher<RelNode> matcher) {
             builder.literal(5));
     assertThat(call.toStringRaw(), is("BETWEEN ASYMMETRIC($0, 1, 5)"));
   }
+
+  @Test public void testSortExchangeFieldTrimmer() {
```
Comment (kgyrtkirk) [inline core/src/test/java/org/apache/calcite/test/RelBuilderTest.java]: I know that we have a lot of catch-all tests - but I think these testcases fall outside of the scope of `RelBuilderTest`

File: core/src/main/java/org/apache/calcite/rel/RelDistributions.java [hunk 2]
```diff
@@ -137,7 +137,7 @@ public RelDistribution apply(Mappings.TargetMapping mapping) {
       return getTraitDef().canonize(
           new RelDistributionImpl(type,
               ImmutableIntList.copyOf(
-                  Mappings.apply((Mapping) mapping, keys))));
```
Comment (kgyrtkirk) [inline core/src/main/java/org/apache/calcite/rel/RelDistributions.java]: why is this change needed? `apply` and `apply2` seems to be different :
* `apply` reorders `list` elements according to `mapping`
** `outList[k]=inList[mappingS[k]]`
* while `apply2` returns the mapping target(s) according to the list
** `outList[k]=mappingT[inList[k]]`

I would feel that in this case `apply` is the right one to use here - why it's not?

File: core/src/test/java/org/apache/calcite/test/RelBuilderTest.java (hunk 1 above)
Comment (kasakrisz) [reply core/src/test/java/org/apache/calcite/test/RelBuilderTest.java]: Moved to `RelFieldTrimmerTest`

File: core/src/main/java/org/apache/calcite/rel/RelDistributions.java (hunk 2 above)
Comment (kgyrtkirk) [reply core/src/main/java/org/apache/calcite/rel/RelDistributions.java]: I agree: at this location we should have `apply2` and not `apply` - it was a bug that `apply` was used here earlier

File: core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    final RelNode root =
+        builder.scan("EMP")
+            .project(builder.field("EMPNO"), builder.field("ENAME"), builder.field("DEPTNO"))
+            .sortExchange(RelDistributions.hash(Lists.newArrayList(1)), RelCollations.EMPTY)
+            .project(builder.field("EMPNO"), builder.field("ENAME"))
+            .build();
+
+    System.out.println(Util.toLinux(RelOptUtil.toString(root)));
+
+    RelFieldTrimmer fieldTrimmer = new RelFieldTrimmer(null, builder);
```
Comment (chunweilei) [inline core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java:107]: No need to print the plan. So I think we'd better remove `System.out.println(Util.toLinux(RelOptUtil.toString(root)));`.

Comment (kasakrisz) [reply core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java:99]: Removed.
